aten/src/ATen/native/GridSampler.cpp

```diff
@@ -0,0 +1,780 @@
+    int64_t grid_sH = grid.stride(1);
+    int64_t grid_sW = grid.stride(2);
+    int64_t grid_sCoor = grid.stride(3);
+    int64_t out_sN = output.stride(0);
+    int64_t out_sC = output.stride(1);
+    int64_t out_sH = output.stride(2);
+    int64_t out_sW = output.stride(3);
+    scalar_t *inp_ptr = input.data<scalar_t>();
+    scalar_t *out_ptr = output.data<scalar_t>();
+    scalar_t *grid_ptr = grid.data<scalar_t>();
+    // loop over each output pixel
+    #ifdef _OPENMP
+    #pragma omp parallel for
+    #endif
+    for (int64_t n = 0; n < N; ++n) {
+      scalar_t *grid_ptr_N = grid_ptr + n * grid_sN;
+      scalar_t *inp_ptr_N = inp_ptr + n * inp_sN;
+      for (int64_t h = 0; h < out_H; ++h) {
+        for (int64_t w = 0; w < out_W; ++w) {
+          // get the corresponding input x, y co-ordinates from grid
+          scalar_t ix = grid_ptr_N[h * grid_sH + w * grid_sW];
+          scalar_t iy = grid_ptr_N[h * grid_sH + w * grid_sW + grid_sCoor];
+
+          // normalize ix, iy from [-1, 1] to [0, inp_W-1] & [0, inp_H-1]
+          ix = ((ix + 1) / 2) * (inp_W - 1);
+          iy = ((iy + 1) / 2) * (inp_H - 1);
+
+          // get NE, NW, SE, SW pixel values from (x, y)
+          int64_t ix_nw = static_cast<int64_t>(std::floor(ix));
+          int64_t iy_nw = static_cast<int64_t>(std::floor(iy));
+          int64_t ix_ne = ix_nw + 1;
+          int64_t iy_ne = iy_nw;
+          int64_t ix_sw = ix_nw;
+          int64_t iy_sw = iy_nw + 1;
+          int64_t ix_se = ix_nw + 1;
+          int64_t iy_se = iy_nw + 1;
+
+          // get surfaces to each neighbor:
+          scalar_t nw = (ix_se - ix)    * (iy_se - iy);
+          scalar_t ne = (ix    - ix_sw) * (iy_sw - iy);
+          scalar_t sw = (ix_ne - ix)    * (iy    - iy_ne);
+          scalar_t se = (ix    - ix_nw) * (iy    - iy_nw);
+
+          if (padding_mode == GridSamplerPadding::Border) {
+            // clip coordinates to image borders
+            ix_nw = clip_coordinates(ix_nw, inp_W);
+            iy_nw = clip_coordinates(iy_nw, inp_H);
+            ix_ne = clip_coordinates(ix_ne, inp_W);
+            iy_ne = clip_coordinates(iy_ne, inp_H);
+            ix_sw = clip_coordinates(ix_sw, inp_W);
+            iy_sw = clip_coordinates(iy_sw, inp_H);
+            ix_se = clip_coordinates(ix_se, inp_W);
+            iy_se = clip_coordinates(iy_se, inp_H);
+          }
+
+          // calculate bilinear weighted pixel value and set output pixel
+          scalar_t *out_ptr_NCHW = out_ptr + n * out_sN + h * out_sH + w * out_sW;
+          scalar_t *inp_ptr_NC = inp_ptr_N;
+          for (int c = 0; c < C; ++c, out_ptr_NCHW += out_sC, inp_ptr_NC += inp_sC) {
+            //   (c, iy_nw, ix_nw) * nw + (c, iy_ne, ix_ne) * ne
+            // + (c, iy_sw, ix_sw) * sw + (c, iy_se, ix_se) * se
+            *out_ptr_NCHW = static_cast<scalar_t>(0);
+            if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_nw, ix_nw, inp_H, inp_W)) {
```

Comment:
I added `padding_mode != GridSamplerPadding::Zeros || ` because the indices can only be out of bounds if padding mode is Zeros. (For Border, the indices are clipped to be within range.)